corgidrp/sorting.py

```diff
@@ -0,0 +1,232 @@
+        cal_frame_list = []
+        n_nonlin = 0
+        nonlin_emgain = []
+        for idx_gain_set, gain_set in enumerate(split_cmdgain[0]):
+            # Frames must be taken consecutively
+            frame_id_list = []
+            exptime_list = []
+            gain_filepath_list = []
+            for frame in gain_set:
+                frame_id_list += [extract_frame_id(frame.filename)]
+                exptime_list += [frame.ext_hdr['EXPTIME']]
+                gain_filepath_list += [frame.filepath]
+            # One can set a stronger condition, though in the end the max set
+            if len(frame_id_list) < 3:
+                continue
+            idx_id_sort = np.argsort(frame_id_list)
+            exptime_arr = np.array(exptime_list)[idx_id_sort]
+            # We need an increasing series of exposure times with the last one
+            # the only repeated value in the series
+            idx_subsets = np.where(np.diff(exptime_arr) < 0)[0]
+            if len(idx_subsets) == 0:
+                continue
+            # length of candidate subsets
+            nonlin_len = []
+            for idx, idx_subset in enumerate(idx_subsets):
+                # Add 0th element plus the one lost in diff
+                if idx == 0:
+                    exptime_tmp = exptime_arr[0:idx_subset+2]
+                else:
+                    exptime_tmp = exptime_arr[idx_subsets[idx-1]+1:idx_subset+2]
+                # Check conditions
+                if (exptime_tmp[-1] in exptime_tmp[:-1] and
+                    len(exptime_tmp) - 1 == len(set(exptime_tmp))):
+                    nonlin_len += [len(exptime_tmp)]
+                else:
+                    nonlin_len += [-1]
+            # COntinue of there are no good candidates
+            if np.max(nonlin_len) <= 0:
+                continue
+            # Find maximum set among good candidates
+            if np.argmax(nonlin_len) == 0:
+                idx_nonlin_first = 0
+                idx_nonlin_last = idx_subsets[0] + 1
+            else:
+                idx_nonlin_first = idx_subsets[np.argmax(nonlin_len)-1] + 1
+                idx_nonlin_last = idx_subsets[np.argmax(nonlin_len)] + 1
+            # Sort unity gain filenames
+            gain_filepath_arr = np.array(gain_filepath_list)[idx_id_sort]
+            cal_list = gain_filepath_arr[idx_nonlin_first:idx_nonlin_last+1]
+            # Update OBSTYPE and take profit to check files are in the list
+            for frame in dataset_cp:
+                if frame.filepath in cal_list:
+                    frame.pri_hdr['OBSTYPE'] = 'NONLIN'
+                    cal_frame_list += [frame]
+                    n_nonlin += 1
+            nonlin_emgain += [split_cmdgain[1][idx_gain_set]]
+
+        print(f'Non-linearity has {n_nonlin} frames with gains', nonlin_emgain)
+        
+    # EM-gain: TODO
+    elif cal_type.lower() == 'em-gain' or cal_type.lower() == 'emgain':
+        print('Considering low EM-gain: TODO')
```

Comment:
shouldn't this be raising an exception?

Reply:
As Marie has mentioned (slack 11/6, 2:08pm): "after chatting with Vanessa, it sounds like we will not need to implement EM-gain sorting after all, which is good news!" And "All that matters is that we correctly select frames from nonlin and kgain, while excluding the frames only useful for EM-gain."

The test function does generate cal frames for EM-gain (vs DAC) and sorts k-gain and non-linearity sets from them. It does not attempt to create a dataset for EM-gain until it becomes a requirement. Let's stick to clear tasks compatible with current specifications in a complex, whose specifications are not very much fixed yet (OBSTYPE example, filename convention, DATETIME, etc) rather than approaching it as a flexible, personal project